File: app/impl/ExamUpdater.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package impl;

import com.google.inject.ImplementedBy;
import java.util.Optional;
import models.assessment.AutoEvaluationConfig;
import models.assessment.ExamFeedbackConfig;
import models.exam.Exam;
import models.user.Role;
import models.user.User;
import play.mvc.Result;

@ImplementedBy(ExamUpdaterImpl.class)
public interface ExamUpdater {
    Optional<Result> updateTemporalFieldsAndValidate(Exam exam, User user, Exam payload);
    Optional<Result> updateStateAndValidate(Exam exam, User user, Exam payload);
    boolean isPermittedToUpdate(Exam exam, User user);
    boolean isAllowedToUpdate(Exam exam, User user);
    boolean isAllowedToRemove(Exam exam);
    void update(Exam exam, Exam payload, Role.Name loginRole);
    void updateAutoEvaluationConfig(Exam exam, AutoEvaluationConfig newConfig);
    Optional<Result> updateLanguage(Exam exam, String code, User user);
    void preparePreview(Exam exam);
    void updateExamFeedbackConfig(Exam exam, ExamFeedbackConfig examFeedbackConfig);
}

File: app/impl/ExamUpdater.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package impl

import com.google.inject.ImplementedBy
import models.assessment.{AutoEvaluationConfig, ExamFeedbackConfig}
import models.exam.Exam
import models.user.{Role, User}
import play.api.mvc.Result

@ImplementedBy(classOf[ExamUpdaterImpl])
trait ExamUpdater:
  def updateTemporalFieldsAndValidate(exam: Exam, user: User, payload: Exam): Option[Result]
  def updateStateAndValidate(exam: Exam, user: User, payload: Exam): Option[Result]
  def isPermittedToUpdate(exam: Exam, user: User): Boolean
  def isAllowedToUpdate(exam: Exam, user: User): Boolean
  def isAllowedToRemove(exam: Exam): Boolean
  def update(exam: Exam, payload: Exam, loginRole: Role.Name): Unit
  def updateAutoEvaluationConfig(exam: Exam, newConfig: AutoEvaluationConfig): Unit
  def updateLanguage(exam: Exam, code: String, user: User): Option[Result]
  def preparePreview(exam: Exam): Unit
  def updateExamFeedbackConfig(exam: Exam, examFeedbackConfig: ExamFeedbackConfig): Unit
